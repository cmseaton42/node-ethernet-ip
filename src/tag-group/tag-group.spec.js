const TagGroup = require("./index");
const Tag = require("../tag");
const Controller = require("../controller");
const { Types } = require("../enip/cip/data-types");

describe("Tag Class", () => {
    describe("Generate Read Requests Method", () => {
        it("Generates Appropriate Output", () => {
            const tag1 = new Tag("helloTag1", "prog", Types.DINT);
            const tag2 = new Tag("helloTag2", "prog", Types.DINT);
            const tag3 = new Tag("helloTag3", "prog", Types.DINT);
            const tag4 = new Tag("helloTag4", "prog", Types.DINT);
            const tag5 = new Tag("helloTag5", "prog", Types.DINT);

            const group = new TagGroup();

            group.add(tag1);
            group.add(tag2);
            group.add(tag3);
            group.add(tag4);
            group.add(tag5);

            expect(group.generateReadMessageRequests()).toMatchSnapshot();
        });

        it("Generates Appropriate Output On Large Data", () => {
            const tag1 = new Tag("Program:Program_Name.Tag_Name1.Member_Name.Another_Member", null, Types.DINT);
            const tag2 = new Tag("Program:Program_Name.Tag_Name2.Member_Name.Another_Member", null, Types.DINT);
            const tag3 = new Tag("Program:Program_Name.Tag_Name3.Member_Name.Another_Member", null, Types.DINT);
            const tag4 = new Tag("Program:Program_Name.Tag_Name4.Member_Name.Another_Member", null, Types.DINT);
            const tag5 = new Tag("Program:Program_Name.Tag_Name5.Member_Name.Another_Member", null, Types.DINT);

            const group = new TagGroup();

            group.add(tag1);
            group.add(tag2);
            group.add(tag3);
            group.add(tag4);
            group.add(tag5);

            expect(group.generateReadMessageRequests()).toMatchSnapshot();
        });
    });

    describe("Generate Write Requests Method", () => {
        it("Generates Appropriate Output", () => {
            const tag1 = new Tag("helloTag1", "prog", Types.DINT);
            const tag2 = new Tag("helloTag2", "prog", Types.DINT);
            const tag3 = new Tag("helloTag3", "prog", Types.DINT);
            const tag4 = new Tag("helloTag4", "prog", Types.DINT);
            const tag5 = new Tag("helloTag5", "prog", Types.DINT);

            const group = new TagGroup();

            group.add(tag1);
            group.add(tag2);
            group.add(tag3);
            group.add(tag4);
            group.add(tag5);

            expect(group.generateWriteMessageRequests()).toMatchSnapshot();

            tag1.value = 0;
            tag2.value = 1;
            tag3.value = 2;
            tag4.value = 3;
            tag5.value = 4;

            group.setController(new Controller());

            expect(group.generateWriteMessageRequests()).toMatchSnapshot();
        });

        it("Generates Appropriate Output On Large Tag Data", () => {
            const tag1 = new Tag("Program:Program_Name.Tag_Name1.Member_Name.Another_Member", null, Types.DINT);
            const tag2 = new Tag("Program:Program_Name.Tag_Name2.Member_Name.Another_Member", null, Types.DINT);
            const tag3 = new Tag("Program:Program_Name.Tag_Name3.Member_Name.Another_Member", null, Types.DINT);
            const tag4 = new Tag("Program:Program_Name.Tag_Name4.Member_Name.Another_Member", null, Types.DINT);
            const tag5 = new Tag("Program:Program_Name.Tag_Name5.Member_Name.Another_Member", null, Types.DINT);

            const group = new TagGroup();

            group.add(tag1);
            group.add(tag2);
            group.add(tag3);
            group.add(tag4);
            group.add(tag5);

            expect(group.generateWriteMessageRequests()).toMatchSnapshot();

            tag1.value = 0;
            tag2.value = 1;
            tag3.value = 2;
            tag4.value = 3;
            tag5.value = 4;

            group.setController(new Controller());

            expect(group.generateWriteMessageRequests()).toMatchSnapshot();
        });
    });

    describe("Length Property", () => {
        it("returns correct length property", () => {
            const tag1 = new Tag("tag1");
            const tag2 = new Tag("tag2");
            const tag3 = new Tag("tag3");
            const tag4 = new Tag("tag4");
            const tag5 = new Tag("tag5");

            const group = new TagGroup();

            expect(group).toHaveLength(0);

            group.add(tag1);
            group.add(tag2);
            group.add(tag3);
            group.add(tag4);
            group.add(tag5);

            expect(group).toHaveLength(5);
        });
    });

    describe("Add Tag Method", () => {
        it("adds tags correctly", () => {
            const tag1 = new Tag("tag");

            const group = new TagGroup();

            expect(group).toHaveLength(0);

            group.add(tag1);

            expect(group).toHaveLength(1);
        });

        it("doesn't add tags more than once", () => {
            const tag1 = new Tag("tag");

            const group = new TagGroup();

            expect(group).toHaveLength(0);

            group.add(tag1);
            group.add(tag1);

            expect(group).toHaveLength(1);
        });
    });

    describe("Remove Tag Method", () => {
        it("removes tags correctly", () => {
            const tag1 = new Tag("tag");

            const group = new TagGroup();

            expect(group).toHaveLength(0);

            group.add(tag1);

            expect(group).toHaveLength(1);

            group.remove(tag1);

            expect(group).toHaveLength(0);
        });

        it("doesn't remove tags that do not exist", () => {
            const tag1 = new Tag("tag1");
            const tag2 = new Tag("tag2");

            const group = new TagGroup();

            expect(group).toHaveLength(0);

            group.add(tag1);

            expect(group).toHaveLength(1);

            group.remove(tag2);

            expect(group).toHaveLength(1);
        });
    });

    describe("Parse Read Messgae Request Method", () => {
        it("reads all tag message data", () => {
            const tag1 = new Tag("tag1", null, Types.DINT);
            const tag2 = new Tag("tag2", null, Types.DINT);
            const tag3 = new Tag("tag3", null, Types.DINT);
            const tag4 = new Tag("tag4", null, Types.DINT);
            const tag5 = new Tag("tag5", null, Types.DINT);

            const group = new TagGroup();

            group.add(tag1);
            group.add(tag2);
            group.add(tag3);
            group.add(tag4);
            group.add(tag5);

            group.setController(new Controller());

            const ids = [];
            const responses = []; 
            let value = 0;
            group.forEach(tag => {
                ids.push(tag.instance_id);
                responses.push({data: Buffer.from(`c4000${value}000000`,"hex")});
                value++;
            });

            group.parseReadMessageResponses(responses,ids);

            expect(tag1.value).toEqual(0);
            expect(tag2.value).toEqual(1);
            expect(tag3.value).toEqual(2);
            expect(tag4.value).toEqual(3);
            expect(tag5.value).toEqual(4);
        });
    });

    describe("Parse Write Messgae Request Method", () => {
        it("unstages wrtie on all tags", () => {
            const tag1 = new Tag("tag1");
            const tag2 = new Tag("tag2");
            const tag3 = new Tag("tag3");
            const tag4 = new Tag("tag4");
            const tag5 = new Tag("tag5");

            const group = new TagGroup();

            group.add(tag1);
            group.add(tag2);
            group.add(tag3);
            group.add(tag4);
            group.add(tag5);

            tag1.value = 0;
            tag2.value = 0;
            tag3.value = 0;
            tag4.value = 0;
            tag5.value = 0;

            expect(tag1.write_ready).toBeTruthy();
            expect(tag2.write_ready).toBeTruthy();
            expect(tag3.write_ready).toBeTruthy();
            expect(tag4.write_ready).toBeTruthy();
            expect(tag5.write_ready).toBeTruthy();

            const ids = [];
            group.forEach(tag => ids.push(tag.instance_id));

            group.parseWriteMessageRequests(null,ids);

            expect(tag1.write_ready).toBeFalsy();
            expect(tag2.write_ready).toBeFalsy();
            expect(tag3.write_ready).toBeFalsy();
            expect(tag4.write_ready).toBeFalsy();
            expect(tag5.write_ready).toBeFalsy();
        });
    });

    describe("Set Controller Method", () => {
        it("sets controller on all tags", () => {
            const tag1 = new Tag("tag1");
            const tag2 = new Tag("tag2");
            const tag3 = new Tag("tag3");
            const tag4 = new Tag("tag4");
            const tag5 = new Tag("tag5");

            const group = new TagGroup();

            group.add(tag1);
            group.add(tag2);
            group.add(tag3);
            group.add(tag4);
            group.add(tag5);

            const plc = new Controller();

            group.setController(plc);

            group.forEach(tag => {
                expect(tag.controller).toMatchObject(plc);
            });
        });
    });
});
