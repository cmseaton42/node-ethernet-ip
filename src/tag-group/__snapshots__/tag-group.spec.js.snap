// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Class Generate Read Requests Method Generates Appropriate Output 1`] = `
Array [
  Object {
    "data": Object {
      "data": Array [
        10,
        2,
        32,
        2,
        36,
        1,
        5,
        0,
        12,
        0,
        42,
        0,
        72,
        0,
        102,
        0,
        132,
        0,
        76,
        13,
        145,
        12,
        80,
        114,
        111,
        103,
        114,
        97,
        109,
        58,
        112,
        114,
        111,
        103,
        145,
        9,
        104,
        101,
        108,
        108,
        111,
        84,
        97,
        103,
        49,
        0,
        1,
        0,
        76,
        13,
        145,
        12,
        80,
        114,
        111,
        103,
        114,
        97,
        109,
        58,
        112,
        114,
        111,
        103,
        145,
        9,
        104,
        101,
        108,
        108,
        111,
        84,
        97,
        103,
        50,
        0,
        1,
        0,
        76,
        13,
        145,
        12,
        80,
        114,
        111,
        103,
        114,
        97,
        109,
        58,
        112,
        114,
        111,
        103,
        145,
        9,
        104,
        101,
        108,
        108,
        111,
        84,
        97,
        103,
        51,
        0,
        1,
        0,
        76,
        13,
        145,
        12,
        80,
        114,
        111,
        103,
        114,
        97,
        109,
        58,
        112,
        114,
        111,
        103,
        145,
        9,
        104,
        101,
        108,
        108,
        111,
        84,
        97,
        103,
        52,
        0,
        1,
        0,
        76,
        13,
        145,
        12,
        80,
        114,
        111,
        103,
        114,
        97,
        109,
        58,
        112,
        114,
        111,
        103,
        145,
        9,
        104,
        101,
        108,
        108,
        111,
        84,
        97,
        103,
        53,
        0,
        1,
        0,
      ],
      "type": "Buffer",
    },
    "tag_ids": Array [
      "f933545900d94fa18e26bf9495c807a5",
      "6a094b0c9172a58a127de4f4e3d07b4d",
      "b76fbe023a62902b18acbd368c5dec70",
      "a7c5544db81347948c6c062509b7664d",
      "a44f360e207bf89c9f2a379caeb459a4",
    ],
  },
]
`;

exports[`Tag Class Generate Write Requests Method Generates Appropriate Output 1`] = `Array []`;
